Sources/LeafKit/LeafLexer.swift

```diff
@@ -1,147 +1,210 @@
+            || isUppercaseLetter
+    }
     
-    init(string: String) {
-        var buffer = ByteBufferAllocator().buffer(capacity: 0)
-        buffer.writeString(string)
-        self.init(template: buffer)
+    var isValidInParameter: Bool {
+        return isLowercaseLetter
+            || isUppercaseLetter
+            || isValidOperator
+            || (.zero ... .nine) ~= self
     }
     
+    var isValidOperator: Bool {
+        switch self {
+        case .plus,
+             .minus,
+             .star,
+             .forwardSlash,
+             .equals,
+             .exclamation,
+             .lessThan,
+             .greaterThan,
+             .ampersand,
+             .vertical:
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+
+struct LeafLexer {
     enum State {
+        // parses as raw, until it finds `#` (excluding escaped `\#`)
         case normal
+        // found a `#`
         case tag
-        case parameters
+        // found a `(` continues until `)`
+        case parameters(depth: Int)
+        // parses a tag body
+        case body
     }
     
     var state: State
     
+    private var buffer: ByteBuffer
+
+    init(string: String) {
+        var buffer = ByteBufferAllocator().buffer(capacity: 0)
+        buffer.writeString(string)
+        self.init(template: buffer)
+    }
+    
     init(template buffer: ByteBuffer) {
-        self.buffer = buffer
         self.state = .normal
+        self.buffer = buffer
     }
     
     mutating func lex() throws -> [LeafToken] {
         var tokens: [LeafToken] = []
-        while let next = try self.next() {
+        while let next = try self.nextToken() {
             tokens.append(next)
         }
         return tokens
     }
     
-    mutating func next() throws -> LeafToken? {
-        guard let next = self.peek() else {
-            // empty
-            return nil
-        }
-        
+    mutating func nextToken() throws -> LeafToken? {
+        guard let next = peek() else { return nil }
         switch state {
         case .normal:
             switch next {
-            case .octothorpe:
-                self.buffer.moveReaderIndex(forwardBy: 1)
-                guard let length = self.countMatching(check: { $0.isAllowedInVariable }) else {
-                    return nil
-                }
-                guard let name = self.buffer.readString(length: length) else {
-                    return nil
+            case .backSlash:
+                // consume '\' only in event of '\#'
+                // otherwise allow it to remain for other
+                // escapes, a la javascript
+                if peek(at: 1) == .octothorpe {
+                    pop()
                 }
-                self.state = .tag
-                return .tag(name: name)
+                // either way, add raw '#' or '\' to registry
+                return buffer.readSlice(length: 1).map(LeafToken.raw)
```

Comment:
i think after consuming the `\` you'd want to return `self.nextToken()` recursively to parse out the tag

Reply:
```swift
if self.peek(at: 1) == .octothorpe {
    self.pop() // remove \
    return self.nextToken() // now parse tag
} else {
    // this is just a normal \
    return buffer.readSlice(length: 1).map(LeafToken.raw) 
}
```

Reply:
we need to stack the `#` key away or the lexer will want to make it a tag in this case 👍